routing/helicopter_router.cpp

```diff
@@ -0,0 +1,139 @@
+  points (1) and (2).
+  Target route should have parameters:
+
+  m_geometry      = [S, S, 1, 1, 2, 2, F, F]
+  m_routeSegments = [S, 1, 1, 2, 2, F, F]
+  m_subroutes     =
+    +-------+--------+-----------------+---------------+
+    | start | finish | beginSegmentIdx | endSegmentIdx |
+    +-------+--------+-----------------+---------------+
+    | S     | 1      | 0               | 2             |
+    | 1     | 1      | 1               | 2             |
+    | 1     | 2      | 2               | 4             |
+    | 2     | 2      | 3               | 4             |
+    | 2     | F      | 4               | 6             |
+    | 2     | F      | 4               | 6             |
+    +-------+--------+-----------------+---------------+
+
+  Constraints:
+  *  m_geometry.size() == 2 * checkpoints.size()
+  *  m_routeSegments.size() == m_geometry.size()-1
+  *  m_subroutes.size() == m_routeSegments.size()-1
+
+ */
+RouterResultCode HelicopterRouter::CalculateRoute(Checkpoints const & checkpoints,
+                                                      m2::PointD const & startDirection,
+                                                      bool adjustToPrevRoute,
+                                                      RouterDelegate const & delegate, Route & route)
+{
+  vector<m2::PointD> const & points = checkpoints.GetPoints();
+  geometry::Altitude const mockAltitude = 0;
+  vector<RouteSegment> routeSegments;
+  vector<double> times;
+  times.reserve(points.size()*2-1);
+  Segment const segment(kFakeNumMwmId, 0, 0, false);
+
+  for (size_t i = 0; i < points.size(); ++i)
+  {
+    turns::TurnItem turn(i, turns::PedestrianDirection::None);
+    geometry::PointWithAltitude const junction(points[i], mockAltitude);
+    RouteSegment::RoadNameInfo const roadNameInfo;
+
+    auto routeSegment = RouteSegment(segment, turn, junction, roadNameInfo);
+    routeSegments.push_back(move(routeSegment));
+    times.push_back(0);
+
+    if (i == points.size() - 1)
+    {
+      // Create final segment.
+      turn = turns::TurnItem(i+1, turns::PedestrianDirection::ReachedYourDestination);
+      RouteSegment lastSegment = RouteSegment(segment, turn, junction, roadNameInfo);
+      routeSegments.push_back(move(lastSegment));
+      times.push_back(0);
+    }
+    else if (i > 0)
+    {
+      // Duplicate intermediate points.
+      RouteSegment intermediateSegment = RouteSegment(segment, turn, junction, roadNameInfo);
+      routeSegments.push_back(move(intermediateSegment));
+      times.push_back(0);
+    }
+  }
+
+  FillSegmentInfo(times, routeSegments);
+  route.SetRouteSegments(move(routeSegments));
+
+  vector<Route::SubrouteAttrs> subroutes;
+  for(size_t i = 1; i < points.size(); ++i)
+  {
+    if (i<points.size()-1)
+    {
+      auto subrt1 = Route::SubrouteAttrs(geometry::PointWithAltitude(points[i-1], mockAltitude),
+                                        geometry::PointWithAltitude(points[i], mockAltitude), i*2-2, i*2);
```

Comment:
In commit [#7566d8f](https://github.com/organicmaps/organicmaps/pull/5381/commits/7566d8fbb8ef1628949f3103deee0d7debc719f2) I increased end index by 1 because in [route.cpp:447](https://github.com/organicmaps/organicmaps/blob/master/routing/route.cpp#L448) we have:

```c++
size_t const segmentIdx = endSegmentIdx - 1;
```

Reply:
It's better to put this comment into the code.

Reply:
@vng PTAL

Reply:
Added comment